### PR TITLE
Error handling

### LINKED ISSUE #213: [FEAT] Implement error codes
- Define error codes for different classes of errors. 
- Make sure errors return information about the location they have been emmited from
- Create verbose error messages.

### PR DESCRIPTION
**Pull request by vhnatyk**
## Describe the changes

finishing of #323 #317 #322 

## Linked Issues

Resolves #213

### CODE REVIEW

File: wrappers/rust/icicle-core/src/msm/mod.rs [hunk 1]
```diff
@@ -138,9 +138,9 @@ macro_rules! impl_msm {
                 points: &[Affine<$curve_config>],
                 cfg: MSMConfig<'a>,
                 results: &mut [Projective<$curve_config>],
-            ) -> CudaResult<()> {
+            ) -> IcicleResult<CudaError> {
                 if points.len() != scalars.len() {
-                    return Err(CudaError::cudaErrorInvalidValue);
+                    panic!("count of points {} and scalars {} do not match:", points.len(), scalars.len());
```
Comment (ChickenLover) [inline wrappers/rust/icicle-core/src/msm/mod.rs:143]: "Number of points" sounds better in my opinion

File: wrappers/rust/icicle-core/src/msm/tests.rs [hunk 2]
```diff
@@ -31,7 +31,7 @@ where
             .stream = &stream;
         cfg.is_async = true;
         cfg.are_results_on_device = true;
-        C::msm(&scalars, &points, cfg, &mut msm_results.as_slice()).unwrap();
+        let _ = C::msm(&scalars, &points, cfg, &mut msm_results.as_slice()).unwrap();
```
Comment (ChickenLover) [inline wrappers/rust/icicle-core/src/msm/tests.rs:34]: Are these `let _ =` really needed in tests?

File: wrappers/rust/icicle-core/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            cuda_error: Some(cuda_error),
+            reason: Some("Runtime CUDA error."),
+        }
+    }
+
+    pub fn from_code_and_reason(icicle_error_code: IcicleErrorCode, reason: &'static str) -> Self {
+        IcicleError {
+            icicle_error_code,
+            reason: Some(reason),
+            cuda_error: Some(CudaError::cudaErrorUnknown),
```
Comment (ChickenLover) [inline wrappers/rust/icicle-core/src/error.rs:47]: Shouldn't it be just `None` if there are no cudaError?

**Review by ChickenLover — COMMENTED** [review]
The Rust code in `dev` is currently formatted, so it would be nice if you run `cargo fmt`

File: wrappers/rust/icicle-core/src/msm/tests.rs (hunk 2 above)
Comment (vhnatyk) [reply wrappers/rust/icicle-core/src/msm/tests.rs]: seems it would be safer - and I got also question from @DmytroTym  - so reverted 👍🏻

**Review by ChickenLover — APPROVED** [review]

**Review by DmytroTym — APPROVED** [review]

**Pull request merged**